content/concepts/peer-routing.md

```diff
@@ -3,6 +3,82 @@ title: Peer Routing
+
+What are `k-buckets`? Remember our initial tree? Here it is. ![Kademlia-system](Kademlia-system.png) See this circled sub-trees? Also, remember that Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
+
+![k-buckets](k-buckets.png)
+
+Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.
+
+## Dynamic DHT
+
+Okay, that was it for static DHT. And the gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
+
+### Leaving
+
+Let us assume that we have some known parameter `k` that represents the number of computers such that it is extremely unlikely that all of those computers will leave the network in the same hour. For example, the authors of the original Kademlia paper experimentally determined that their `k` should be 20.
+
+Next, our `k-bucket` should, instead of storing just 1 computer, store `k` computers within that `k-bucket’s` range. Thus, with high likelihood, there will always be at least one computer online in each of a computer’s `k-buckets`.
+
+And, when hopping, instead of asking one computer from `k-bucket` we will be asking `k`. Some may not reply, but not all of them.
+
+To check if node is still alive we can ping it periodically. If node doesn't respond in the allotted time it will be marked as non-responding and eventually replaced once we have a better one.
+
+How do we become aware of new computers that could fit into our unfilled `k-buckets`? We could perform lookups on random ids within a `k-bucket’s` id range (the range is defined by all leaves that are direct descendants of the `k-bucket’s` corresponding internal node, which is not leaf, in the complete tree) and thus learn about computers within that `k-bucket` if they exist. This is called a _bucket refresh_.
+
+### Joining
+
+To join the DHT new computer `Eve` must know at least on IP address of computer which is in the network. Then `Eve` will perform a lookup of herself to get closest known computers which are already aware of `Eve`. And after that `Eve` will perform a _bucket refresh_ for all `k-buckets` from the closest known computers to `Eve`. This will populate the `Eve's` routing table based on the correctness of the lookup algorithm and the fact that the routing tables of other computers are sufficiently populated. In addition, all computers that are being queried by `Eve` will become aware of `Eve` in the process and thus will have the opportunity to insert `Eve` into their own `k-buckets` if their corresponding bucket is not already full.
```

Comment:
```suggestion
To join the DHT new computer, which we will call `Eve`,  must know at least on IP address of a computer which is in the network. Then `Eve` will perform a lookup of herself to get closest known computers which are already aware of `Eve`. After that `Eve` will perform a _bucket refresh_ for all `k-buckets` from the closest known computers to `Eve`. 

This will populate the `Eve's` routing table based on the correctness of the lookup algorithm and the fact that the routing tables of other computers are sufficiently populated. In addition, all computers that are being queried by `Eve` will become aware of `Eve` in the process and thus will have the opportunity to insert `Eve` into their own `k-buckets` if their corresponding bucket is not already full.
```